### DISCUSSION TITLE
How to deploy NiceGUI application to Heroku?

### DISCUSSION THREAD

**Discussion by 200729** [zauberzeug/nicegui#25] [Q&A] [answered]
I tried to deploy hellord app on Heroku:
code:

```
from nicegui import ui
import os

ui.label('Hello NiceGUI!')
ui.button('BUTTON', on_click=lambda: print('button was pressed', flush=True))

port = int(os.getenv('PORT'))
ui.run(port=port)
```

and I have error:

`Exception: Could not find ui.run() command`

Is possible to deploy on Heroku?

**Comment by falkoschindler** [comment]
Thanks for your feedback! You spotted an issue in NiceGUI's pre-evaluation of the `ui.run` command. Already during import we try to find this command at the end of the main script in order to tell whether the server should start with auto-reload or not. This works well in most cases. But when the specified port (or any other argument) is a variable or an even more complex expression, our pre-evaluation fails.

But we can avoid terminating the whole app and simply continue evaluating the script in a so-called interactive mode (same as if NiceGUI is used in an interactive Python shell). The auto-reloading feature does not work here, but at least the UI starts as expected and whatever expression is passed to `ui.run` will be evaluated correctly.

Please have a try with version 0.7.13. I haven't tested it on Heroku, yet. But I'm pretty confident that this new release will solve your issue.

**Reply by 200729** [reply]
Thanks for quick answer and fix! I tried version 0.7.13 and error `Exception: Could not find ui.run() command` is not visible. However as it is in life of a programmer fixing one error discovers another error :D Now there is a warning, but application crashes. Warning seems to be from justpy:
`WARNING:  You must pass the application as an import string to enable 'reload' or 'workers'.`
Do you have idea how to fix it?

**Reply by falkoschindler** [reply]
Oh, that's interesting. Would you mind sharing your [deployment setup](https://bocadilloproject.github.io/how-to/heroku.html#setting-things-up) like Procfile, requirements.txt and runtime.txt? I could imagine that Heroku has trouble finding the right entry point for the guvicorn server.

**Reply by 200729** [reply]
Sure, here is whole hello world project in zip [nicegui_tets.zip](https://github.com/zauberzeug/nicegui/files/8230247/nicegui_tets.zip)

**Reply by falkoschindler** [reply]
Thanks! I identified two issues, one regarding your Heroku Settings and one deep inside the NiceGUI configuration... I'm still working on it. 🤞

**Reply by falkoschindler** [reply]
With version 0.7.15 it finally seems to work as expected. See [my other answer](https://github.com/zauberzeug/nicegui/discussions/25#discussioncomment-2349263) for more details.

**Comment by falkoschindler** [comment, marked as answer]
The new NiceGUI version 0.7.15 should finally work nicely on Heroku. The main issue was that the environment variable `PORT` has been overwritten by the import statement `from nicegui import ui`. But now the variables `HOST` and `PORT` - if defined - will be read and used during `ui.run()`. So you don't even need to read them via `os.getenv` and pass them to `ui.run`.

In the meantime I thought using Heroku's default setting of `WEB_CONCURRENCY=2` might be an issue. So I experimented with [setting it to 1](https://devcenter.heroku.com/articles/config-vars#using-the-heroku-dashboard). But with version 0.7.15 this step is probably not necessary anymore.

So my final setup is as follows:

**main.py**
```python
from nicegui import ui

label = ui.label('Hello NiceGUI!')
ui.button('BUTTON', on_click=lambda: label.set_text(label.text + '!'))

ui.run()
```

**Procfile**
```
web: python main.py
```

**requirements.txt**
```
nicegui==0.7.15
```

**runtime.txt**
```
python-3.9.10
```

**Reply by 200729** [reply]
Great! Thank you! It is working now.

**Reply by alikalik9** [reply]
Hi, I am trying to deploy to Heroku using the same set up and I am getting the following error message:
Error R10 (Boot timeout) -> Web process failed to bind to $PORT within 60 seconds of launch

**Reply by falkoschindler** [reply]
Hi @alikalik9! I'm not sure about this Error R10. We are currently working on another problem with Heroku (#469).
@rodja Do you think this could be related? Either way, I guess we need some documentation about hosting with Heroku.

**Reply by alikalik9** [reply]
Thanks for the quick answer. Yes, a documentation for Heroku (or any other convinient provider) would be much appteciated!
Keep up the good work!
